compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4851,8 +4936,226 @@ public void visit(BLangTypedescExpr accessExpr, AnalyzerData data) {
         data.resultType = types.checkType(accessExpr, actualType, data.expType);
     }
 
+    public LinkedHashSet<BType> getBasicNumericTypes(LinkedHashSet<BType> memberTypes) {
+        LinkedHashSet<BType> basicNumericTypes = new LinkedHashSet<>(memberTypes.size());
+
+        for (BType value : memberTypes) {
+            BType referredType = types.getReferredType(value);
+            int typeTag = referredType.tag;
+            if (TypeTags.isIntegerTypeTag(typeTag)) {
+                basicNumericTypes.add(symTable.intType);
+            } else if (typeTag == TypeTags.FLOAT || typeTag == TypeTags.DECIMAL) {
+                basicNumericTypes.add(value);
+            } else if (typeTag == TypeTags.JSON || typeTag == TypeTags.ANYDATA || typeTag == TypeTags.ANY) {
+                basicNumericTypes.add(symTable.intType);
+                basicNumericTypes.add(symTable.floatType);
+                basicNumericTypes.add(symTable.decimalType);
+                break;
+            } else if (typeTag == TypeTags.FINITE) {
+                LinkedHashSet<BType> typesInValueSpace = getTypesInFiniteValueSpace((BFiniteType) referredType);
+                basicNumericTypes.addAll(getBasicNumericTypes(typesInValueSpace));
+            }
+        }
+        return basicNumericTypes;
+    }
+
+    public BType createFiniteTypeForNumericUnaryExpr(BLangUnaryExpr unaryExpr, AnalyzerData data) {
+        BLangNumericLiteral newNumericLiteral = Types.constructNumericLiteralFromUnaryExpr(unaryExpr);
+        BTypeSymbol finiteTypeSymbol = Symbols.createTypeSymbol(SymTag.FINITE_TYPE,
+                0, Names.EMPTY, data.env.enclPkg.symbol.pkgID, null, data.env.scope.owner,
+                unaryExpr.pos, SOURCE);
+        BFiniteType finiteType = new BFiniteType(finiteTypeSymbol);
+        finiteType.addValue(newNumericLiteral);
+        finiteTypeSymbol.type = finiteType;
+
+        types.setImplicitCastExpr(unaryExpr, unaryExpr.expr.getBType(), data.expType);
+        return finiteType;
+    }
+
+    public LinkedHashSet<BType> getTypesInFiniteValueSpace(BFiniteType referredType) {
+        Set<BLangExpression> valueSpace = referredType.getValueSpace();
+        LinkedHashSet<BType> typesInValueSpace = new LinkedHashSet<>(valueSpace.size());
+        for (BLangExpression expr : valueSpace) {
+            typesInValueSpace.add(expr.getBType());
+        }
+        return typesInValueSpace;
+    }
+
+    public BType getNewExpectedTypeForFiniteAndUnion(LinkedHashSet<BType> numericTypes, BType newExpectedType) {
+        LinkedHashSet<BType> basicNumericTypes = getBasicNumericTypes(numericTypes);
```

Comment:
What's the reason to add this new applicable preferred type logic? Don't we have existing logic for numeric literal compatibility? cc @chiranSachintha

Reply:
With the current logic, for ex: `-1f|-1|-1d|"string" x = -1; ` the type of -1 comes as `int`, which is the expected result. If we replace `getNewExpectedTypeForFiniteAndUnion` with `getPreferredMemberTypeTag` we get `float` as the preferred type, which is not correct for this case